File: app/src/main/java/com/example/android/popularmovies/model/Movie.java
package com.example.android.popularmovies.model;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

//ToDo 9: Now If you run , you will see that the list not order by popularity
// we will add popularity to our database and migrate it

@Entity(tableName = "movie_table")
public class Movie implements Parcelable {

    private static final String IMAGE_BASE_URL = "https://image.tmdb.org/t/p/w500";

    @PrimaryKey
    @NonNull
    @SerializedName("id")
    private Integer id;

    @SerializedName("title")
    private String title;

    @SerializedName("poster_path")
    private String posterPath;

    @SerializedName("backdrop_path")
    private String backdropPath;

    @SerializedName("release_date")
    private String releaseDate;

    @SerializedName("overview")
    private String overview;

    @SerializedName("vote_average")
    private Double voteAverage;


    @Ignore
    @SerializedName("adult")
    private boolean adult;
    @Ignore
    @SerializedName("genre_ids")
    private List<Integer> genreIds = new ArrayList<>();
    @Ignore
    @SerializedName("original_title")
    private String originalTitle;
    @Ignore
    @SerializedName("original_language")
    private String originalLanguage;
    //@Ignore //ToDo 9.1: not ignore and add getter and setter
    @SerializedName("popularity")
    private Double popularity;
    @Ignore
    @SerializedName("vote_count")
    private Integer voteCount;
    @Ignore
    @SerializedName("video")
    private Boolean video;


    public Movie() {

    }


    public Movie(String posterPath, boolean adult, String overview, String releaseDate,
                 List<Integer> genreIds, @NonNull Integer id, String originalTitle, String originalLanguage,
                 String title, String backdropPath, Double popularity,
                 Integer voteCount, Boolean video, Double voteAverage) {
        this.posterPath = posterPath;
        this.adult = adult;
        this.overview = overview;
        this.releaseDate = releaseDate;
        this.genreIds = genreIds;
        this.id = id;
        this.originalTitle = originalTitle;
        this.originalLanguage = originalLanguage;
        this.title = title;
        this.backdropPath = backdropPath;
        this.popularity = popularity;
        this.voteCount = voteCount;
        this.video = video;
        this.voteAverage = voteAverage;
    }

    protected Movie(Parcel in) {
        posterPath = in.readString();
        adult = in.readByte() != 0;
        overview = in.readString();
        releaseDate = in.readString();
        if (in.readByte() == 0) {
            id = null;
        } else {
            id = in.readInt();
        }
        originalTitle = in.readString();
        originalLanguage = in.readString();
        title = in.readString();
        backdropPath = in.readString();
        if (in.readByte() == 0) {
            popularity = null;
        } else {
            popularity = in.readDouble();
        }
        if (in.readByte() == 0) {
            voteCount = null;
        } else {
            voteCount = in.readInt();
        }
        byte tmpVideo = in.readByte();
        video = tmpVideo == 0 ? null : tmpVideo == 1;
        if (in.readByte() == 0) {
            voteAverage = null;
        } else {
            voteAverage = in.readDouble();
        }
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(posterPath);
        dest.writeByte((byte) (adult ? 1 : 0));
        dest.writeString(overview);
        dest.writeString(releaseDate);
        dest.writeByte((byte) 1);
        dest.writeInt(id);
        dest.writeString(originalTitle);
        dest.writeString(originalLanguage);
        dest.writeString(title);
        dest.writeString(backdropPath);
        if (popularity == null) {
            dest.writeByte((byte) 0);
        } else {
            dest.writeByte((byte) 1);
            dest.writeDouble(popularity);
        }
        if (voteCount == null) {
            dest.writeByte((byte) 0);
        } else {
            dest.writeByte((byte) 1);
            dest.writeInt(voteCount);
        }
        dest.writeByte((byte) (video == null ? 0 : video ? 1 : 2));
        if (voteAverage == null) {
            dest.writeByte((byte) 0);
        } else {
            dest.writeByte((byte) 1);
            dest.writeDouble(voteAverage);
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Movie> CREATOR = new Creator<Movie>() {
        @Override
        public Movie createFromParcel(Parcel in) {
            return new Movie(in);
        }

        @Override
        public Movie[] newArray(int size) {
            return new Movie[size];
        }
    };


    public String getPosterPath() {
        if (posterPath != null && !posterPath.contains(IMAGE_BASE_URL))
            return IMAGE_BASE_URL + posterPath;
        else
            return posterPath;
    }

    public String getOverview() {
        return overview;
    }

    public String getReleaseDate() {
        return releaseDate;
    }

    public List<Integer> getGenreIds() {
        return genreIds;
    }

    public Integer getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getBackdropPath() {

        if (backdropPath != null && !backdropPath.contains(IMAGE_BASE_URL))
            return IMAGE_BASE_URL + backdropPath;
        else
            return backdropPath;

    }

    public Double getVoteAverage() {
        return voteAverage;
    }


    public void setId(@NonNull Integer id) {
        this.id = id;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setPosterPath(String posterPath) {
        this.posterPath = posterPath;
    }

    public void setBackdropPath(String backdropPath) {
        this.backdropPath = backdropPath;
    }

    public void setReleaseDate(String releaseDate) {
        this.releaseDate = releaseDate;
    }

    public void setOverview(String overview) {
        this.overview = overview;
    }

    public void setVoteAverage(Double voteAverage) {
        this.voteAverage = voteAverage;
    }

    public Double getPopularity() {
        return popularity;
    }

    public void setPopularity(Double popularity) {
        this.popularity = popularity;
    }
}

File: app/src/main/kotlin/com/example/android/popularmovies/model/Movie.kt
package com.example.android.popularmovies.model

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName


class Movie() : Parcelable {

    private val imageBaseUrl = "https://image.tmdb.org/t/p/w500"

    @SerializedName("poster_path")
    var posterPath: String? = null
        get() {
            return when {
                field == null -> return ""
                !field?.contains(imageBaseUrl)!! -> imageBaseUrl + field
                else -> field
            }
        }

    @SerializedName("overview")
    var overview: String? = null

    @SerializedName("release_date")
    var releaseDate: String? = null

    @SerializedName("id")
    var id: Int? = null

    @SerializedName("title")
    var title: String? = null

    @SerializedName("vote_count")
    var voteCount: Int? = null

    @SerializedName("vote_average")
    var voteAverage: Float? = null

    @SerializedName("backdrop_path")
    var backdropPath: String? = null
        get() {
            return when {
                field == null -> return ""
                !field?.contains(imageBaseUrl)!! -> imageBaseUrl + field
                else -> field
            }
        }

    constructor(parcel: Parcel) : this() {
        posterPath = parcel.readString()
        overview = parcel.readString()
        releaseDate = parcel.readString()
        id = parcel.readValue(Int::class.java.classLoader) as? Int
        title = parcel.readString()
        voteCount = parcel.readValue(Int::class.java.classLoader) as? Int
        voteAverage = parcel.readValue(Float::class.java.classLoader) as? Float
        backdropPath = parcel.readString()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(posterPath)
        parcel.writeString(overview)
        parcel.writeString(releaseDate)
        parcel.writeValue(id)
        parcel.writeString(title)
        parcel.writeValue(voteCount)
        parcel.writeValue(voteAverage)
        parcel.writeString(backdropPath)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Movie> {
        override fun createFromParcel(parcel: Parcel): Movie {
            return Movie(parcel)
        }

        override fun newArray(size: Int): Array<Movie?> {
            return arrayOfNulls(size)
        }
    }


}